kong/router/fields.lua

```diff
@@ -0,0 +1,236 @@
+    end,
+
+    ["http.host"] =
+    function(v, params, cb)
+      return cb("http.host", params.host)
+    end,
+
+    -- tls.*
+
+    ["tls.sni"] =
+    function(v, params, cb)
+      return cb("tls.sni", params.sni)
+    end,
+
+    -- net.*
+
+    ["net.src.ip"] =
+    function(v, params, cb)
+      return cb("net.src.ip", params.src_ip)
+    end,
+
+    ["net.src.port"] =
+    function(v, params, cb)
+      return cb("net.src.port", params.src_port)
+    end,
+
+    ["net.dst.ip"] =
+    function(v, params, cb)
+      return cb("net.dst.ip", params.dst_ip)
+    end,
+
+    ["net.dst.port"] =
+    function(v, params, cb)
+      return cb("net.dst.port", params.dst_port)
+    end,
+
+    -- below are atc context only
+
+    ["net.protocol"] =
+    function(v, params, cb)
+      return cb("net.protocol", params.scheme)
+    end,
+
+    ["net.port"] =
+    function(v, params, cb)
+      return cb("net.port", params.port)
+    end,
+}
+
+
+local COMPLEX_FIELDS_FUNCS = {
+    ["http.headers."] =
+    function(v, params, cb)
+      local headers = params.headers
+      if not headers then
+        return true
+      end
+
+      for _, name in ipairs(v) do
+        local value = headers[name]
+
+        local res, err = cb("http.headers." .. name, value,
+                            replace_dashes_lower) -- only for cache_key
+        if not res then
+          return nil, err
+        end
+      end -- for ipairs(v)
+
+      return true
+    end,
+
+    ["http.queries."] =
+    function(v, params, cb)
+      local queries = params.queries
+      if not queries then
+        return true
+      end
+
+      for _, name in ipairs(v) do
+        local value = queries[name]
+
+        local res, err = cb("http.queries." .. name, value)
+        if not res then
+          return nil, err
+        end
+      end -- for ipairs(v)
+
+      return true
+    end,
+}
+
+
+-- cache key string
+local str_buf = buffer.new(64)
+
+
+local function get_cache_key(fields, params)
+  str_buf:reset()
+
+  -- NOTE: DO NOT yield until str_buf:get()
+  for field, value in pairs(fields) do
+
+    -- these fields were not in cache key
+    if field == "net.protocol" or
+       field == "net.port"
+    then
+      goto continue
+    end
+
+    local func = SIMPLE_FIELDS_FUNCS[field]
+
+    if func then
+      func(value, params, function(field, value)
+        str_buf:put(value or ""):put("|")
+        return true
+      end)
+
+      goto continue
+    end -- if func
+
+    func = COMPLEX_FIELDS_FUNCS[field]
+
+    if func then  -- http.headers.* or http.queries.*
+      func(value, params, function(field, value, lower_func)
+        if lower_func then
+          field = lower_func(field)
+        end
+
+        if type(value) == "table" then
+          tb_sort(value)
+          value = tb_concat(value, ",")
+        end
+
+        str_buf:putf("%s=%s|", field, value or "")
+
+        return true
+      end)
+    end -- if func
+
+    -- ignore unknown fields
```

Comment:
Unknown field should not happen normally, better assert here or return an error instead.

Reply:
done